### ISSUE TITLE
Mktemp dependency is neither POSIX nor Windows compatible

### ISSUE THREAD

**Issue by kotmage** [OpenVPN/easy-rsa#84]
Source of problem:
https://github.com/OpenVPN/easy-rsa/pull/26 Good intentions, but we need to revert it or make i dummy mktemp.

**Comment by cbridgerg00** [comment]
Just to add to this: I was trying to run easy-rsa on a Ubiquiti EdgeRouter Lite, which I believe is running Debian Wheezy, and I got the error "mktemp: invalid option -- 'u'"

I understand that the -u option is used to just retrieve the filename, but I guess on Wheezy mktemp does not have a -u option. I was able to resolve by editing the script to remove the -u option, but I am not sure if that will break the script (it at least ran without errors).

Anyway, I assume that is a problem with mktemp not being POSIX compatible, so I am leaving this here instead of creating a new issue (seemed more appropriate).

```
:/etc/openvpn/easy-rsa/EasyRSA-3.0.1$ sudo ./easyrsa build-ca
mktemp: invalid option -- 'u'
BusyBox v1.19.0 (2014-10-09 14:06:57 PDT) multi-call binary.

Usage: mktemp [-dt] [-p DIR] [TEMPLATE]

Create a temporary file with name based on TEMPLATE and print its name.
TEMPLATE must end with XXXXXX (e.g. [/dir/]nameXXXXXX).
Without TEMPLATE, -t tmp.XXXXXX is assumed.

        -d      Make directory, not file
        -t      Prepend base directory name to TEMPLATE
        -p DIR  Use DIR as a base directory (implies -t)

Base directory is: -p DIR, else $TMPDIR, else /tmp

mktemp: invalid option -- 'u'
BusyBox v1.19.0 (2014-10-09 14:06:57 PDT) multi-call binary.

Usage: mktemp [-dt] [-p DIR] [TEMPLATE]

Create a temporary file with name based on TEMPLATE and print its name.
TEMPLATE must end with XXXXXX (e.g. [/dir/]nameXXXXXX).
Without TEMPLATE, -t tmp.XXXXXX is assumed.

        -d      Make directory, not file
        -t      Prepend base directory name to TEMPLATE
        -p DIR  Use DIR as a base directory (implies -t)

Base directory is: -p DIR, else $TMPDIR, else /tmp

Generating a 2048 bit RSA private key
..................+++
...............................+++
writing new private key to ''
No such file or directory
2006533272:error:02001002:system library:fopen:No such file or directory:bss_fil           e.c:398:fopen('','w')
2006533272:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:400:

Easy-RSA error:

Failed to build the CA
```

**Comment by bagodoosh** [comment]
petty fix -- modify easyrsa script and insert the following after print() line 223:
(location is not essential as long as it's before build_ca() line 429. don't insert it inside another function. if you're not familiar with shell scripts then stick to the instruction above)

```
mktemp() {
 # ignore all parameters except -u
 us=0			#unsafe
 while [ "${1:0:1}" == "-" ]; do
  [ "$1" == "-u" ] && us=1
  shift
 done
 xt=${1##*.}	#extention
 bn=${1%.*}	#basename
 ln=${#xt}	#ext length
 fn=""
 until [ ! -f $fn ]; do
  xm=$$$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM
  xn=${xm:0:ln}	#new extention with PID & random #s appended up to length of ext
  fn=${bn}.${xn}
 done
 [ $us -eq 0 ] && echo -n >$fn	#touch new temp file
 echo $fn
}
```

PS> i think newer versions of mksh implement mktemp; however mksh/w32 is a beta version of r39 from 2013.

**Comment by mark4asp** [comment]
"fabulous fix", not. It's like going back in time to BASIC.  I get an infinite loop telling me:

./easyrsa[1236]: [: Files/OpenVPN/easyRSA-3.0.1/pki/private/ca.key.1181299942: unexpected operator/operand

Nice try. I shall attempt to hunt down a version of mksh with mktemp

**Comment by bagodoosh** [comment]
at least you're more proficient at sarcasm than technical know-how. i think you goofed at copy/pasting. it's reaching the end of the script unexpectedly. did you omit the close brace "}"?
this is how mine looks like (in notepad++):

![image](https://user-images.githubusercontent.com/14930037/29287771-94ab969c-8104-11e7-839f-ae5f6018bcd1.png)

**Comment by ecrist** [comment]
This has been fixed in v3.0.3.

**Comment by ammgws** [comment]
I can't see a fix in either master nor v3.0.3. "mktemp -u" is still called throughout the script which fails on Busybox as noted in the second comment.
